### PR TITLE
PA-926 Enable various eslint rules

### PR DESCRIPTION
**Pull request by yonran**
[PA-926](https://mavencorp.atlassian.net/browse/PA-926) Add additional recommended rules to eslint-maven that can catch common mistakes

Enable a subset of `eslint:recommended`, `plugin:@typescript-eslint/recommended`, and `plugin:@typescript-eslint/recommended-requiring-type-checking`

Enable a high-precision subset of eslint:recommended that have few false positives. There are other useful rules (e.g. no-unused-vars, no-undef) but they probably require more effort to enable

Enable several recommended rules from @typescript-eslint:

Enable no-floating-promises, which finds statements that might result in UnhandledPromiseRejectionWarning. Usually these promises should be awaited so that the error will be propagated to the caller. If you don’t want to await it, you can explicitly wrap it in void().

Enable no-misused-promises, which identifies promises used in conditionals; they should be awaited first. It also identifies async functions that are used where functions returning void are expected; they should be converted to non-async functions that wrap the async function.

Enable no-extra-non-null-assertion, no-for-in-array, no-misused-new, no-non-null-asserted-optional-chain, which identify common mistakes with few false positives.

Enable no-namespace, prefer-namespace-keyword, which discourage deprecated TypeScript features

Other recommended rules are more opinionated (no-empty-interface, no-inferrable-types, no-unnecessary-type-assertion, no-unnecessary-type-constraint, prefer-as-const) or difficult to enable (no-explicit-any, no-non-null-assertion, no-this-alias, no-unsafe-argument, no-unsafe-assignment, no-unsafe-call, no-unsafe-member-access, no-unsafe-return, restrict-plus-operands, restrict-template-expressions, triple-slash-reference, unbound-method). Please consider extending plugin:@typescript-eslint/recommended and plugin:@typescript-eslint/recommended-requiring-type-checking on new projects.

## Should eslint-config-maven contain more aggressive rules?

Ideally, I think we should enable more of `eslint:recommended`, `plugin:@typescript-eslint/recommended`, and `plugin:@typescript-eslint/recommended-requiring-type-checking`. But they would require more effort. In particular:
Here are useful rules from [eslint:recommended](https://eslint.org/docs/rules/) that I did *not* enable because they would have a lot of code changes. But in greenfield development we should probably enable them:
* [no-undef](https://eslint.org/docs/rules/no-undef) disallow the use of undeclared variables unless mentioned in `/*global */` comments
* [no-unreachable](https://eslint.org/docs/rules/no-unreachable) disallow unreachable code after `return`, `throw`, `continue`, and `break` statements
* [no-unused-vars](https://eslint.org/docs/rules/no-unused-vars) disallow unused variables

Here are useful rules from [@typescript-eslint](https://github.com/typescript-eslint/typescript-eslint/tree/v5.15.0/packages/eslint-plugin/docs/rules#supported-rules)/`recommended` or `recommended-requiring-type-checking` that I did *not* enable because they would have a lot of code changes. But in greenfield development we should probably enable them:
* no-explicit-any Disallow usage of the `any` type
* no-non-null-assertion Disallows non-null assertions using the `!` postfix operator
* no-this-alias Disallow aliasing `this`
* no-unsafe-argument Disallows calling a function with an `any` type value
* no-unsafe-assignment Disallows assigning `any` to variables and properties
* no-unsafe-call Disallows calling an `any` type value
* no-unsafe-member-access Disallows member access on `any` typed variables
* no-unsafe-return Disallows returning `any` from a function
* restrict-plus-operands When adding two variables, operands must both be of type number or of type string
* restrict-template-expressions Enforce template literal expressions to be of string type
* triple-slash-reference Sets preference level for triple slash directives versus ES6-style import declarations
* unbound-method Enforces unbound methods are called with their expected scope

### CODE REVIEW

**Comment by martiantim** [comment]
I don't have strong opinions on the specifics so I added other reviewers from the team that might :)

@yonran please attach to this ticket for tracking: https://mavencorp.atlassian.net/browse/PA-926

**Review by robmaven — APPROVED** [review]
Anything that can lead to more consistent code across projects is good by me. The more rules, the better!

**Review by stanimoto — APPROVED** [review]
> Anything that can lead to more consistent code across projects is good by me. The more rules, the better!

I second that!

We may want to add those rules that should be enabled in greenfield dev to https://github.com/themaven-net/project-template/blob/master/tsconfig.json for now?

**Comment by yonran** [comment]
@robmaven I added support for linting both js files and ts files depending on the file type, and I added support to automatically publish to GitHub Packages when a tag is pushed.

File: eslint-js.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        'spaced-comment': 'off',
+        'no-useless-escape': 'off', // TODO should probably be on, with our errors fixed
+        'prefer-promise-reject-errors': 'off', // TODO consider fixing
+        'no-unneeded-ternary': 'off', // TODO should probably be on, with our errors fixed
+        'no-useless-return': 'off',
+        'new-cap': 'off',
+        'no-path-concat': 'off',
+        'no-throw-literal': 'off', // TODO should probably be on, with our errors fixed
+        'no-template-curly-in-string': 'off',
+        yoda: 'off',
```
Comment (robmaven) [inline eslint-js.js:56]: had to look up this one 😀

**Comment by robmaven** [comment]
LGTM!

**Pull request merged**